pageserver/src/keyspace.rs

```diff
@@ -61,6 +62,61 @@ impl KeySpace {
 
         KeyPartitioning { parts }
     }
+
+    /// Add range to keyspace.
+    ///
+    /// Unlike KeySpaceAccum, it accepts key ranges in any order and overlapping ranges.
+    pub fn add_range(&mut self, range: Range<Key>) {
+        let start = range.start;
+        let mut end = range.end;
+        let mut prev_index = match self.ranges.binary_search_by_key(&end, |r| r.start) {
+            Ok(index) => index,
+            Err(0) => {
+                self.ranges.insert(0, range);
+                return;
+            }
+            Err(index) => index - 1,
+        };
+        loop {
+            let mut prev = &mut self.ranges[prev_index];
+            if prev.end >= start {
+                // two ranges overlap
+                if prev.start <= start {
+                    // combine with prev range
+                    if prev.end < end {
+                        prev.end = end;
+                        debug!("Extend wanted image {}..{}", prev.start, end);
+                    }
+                    return;
+                } else {
+                    if prev.end > end {
+                        end = prev.end;
+                    }
+                    self.ranges.remove(prev_index);
+                }
+            } else {
+                prev_index += 1; // insert after prev range
+                break;
+            }
+            if prev_index == 0 {
+                break;
+            }
+            prev_index -= 1;
+        }
+        debug!("Wanted image {}..{}", start, end);
+        self.ranges.insert(prev_index, start..end);
+    }
+
+    ///
+    /// Check if key space contains overlapping range
+    ///
+    pub fn overlaps(&self, range: &Range<Key>) -> bool {
+        match self.ranges.binary_search_by_key(&range.end, |r| r.start) {
+            Ok(_) => false,
+            Err(0) => false,
+            Err(index) => self.ranges[index - 1].end > range.start,
+        }
+    }
```

Comment:
I wrote some unit tests for `overlaps` too, and found one bug:
```

    #[test]
    fn keyspace_overlaps() {
        let mut ks = KeySpace::default();
        ks.add_range(kr(10..20));
        ks.add_range(kr(30..40));

        //        #####      #####
        // xxxx
        assert!(ks.overlaps(&kr(0..5)) == false);
        
        //        #####      #####
        //   xxxx
        assert!(ks.overlaps(&kr(5..9)) == false);

        //        #####      #####
        //    xxxx
        assert!(ks.overlaps(&kr(5..10)) == false);


        //        #####      #####
        //     xxxx
        assert!(ks.overlaps(&kr(5..11)) == true);
        
        //        #####      #####
        //        xxxx
        assert!(ks.overlaps(&kr(10..15)) == true);

        //        #####      #####
        //         xxxx
        assert!(ks.overlaps(&kr(15..20)) == true);

        //        #####      #####
        //           xxxx
        assert!(ks.overlaps(&kr(15..25)) == true);

        //        #####      #####
        //              xxxx
        assert!(ks.overlaps(&kr(22..28)) == false);

        //        #####      #####
        //               xxxx
        assert!(ks.overlaps(&kr(25..30)) == false);

        //        #####      #####
        //                      xxxx
        assert!(ks.overlaps(&kr(35..35)) == true);

        //        #####      #####
        //                        xxxx
        assert!(ks.overlaps(&kr(40..45)) == false);

        //        #####      #####
        //                        xxxx
        assert!(ks.overlaps(&kr(45..50)) == false);

        //        #####      #####
        //        xxxxxxxxxxx
        assert!(ks.overlaps(&kr(0..30)) == true);  // XXXXX This fails currently!
    }
```

Reply:
It is bug not in `add_range` but in `overlaps`
Fixed